scripts/create_singularities

```diff
@@ -40,264 +32,279 @@
+        last_version = self.get_last_docker_version_tag(dockerhubid)
+        if last_version is not None:
+            last_version_pure = re.sub(r"@.*\Z", "", last_version)
+            last_version_tag = re.sub(r"^[^@]*@", "", last_version)
+            # print(dockerhubid, last_version)
+            singfile = self.create_singularity_file(
+                dockerhubid, last_version_tag, last_version_pure, family
+            )
+            self.add_singularity_versioned_image(
+                singfile, "ReproNim/containers", family, dockerhubid
+            )
+
+
+@dataclass
+class SingularityFile:
+    sourcefile: Path
+    imagename: str
+
+    def build(self, builder: Builder) -> Optional[Path]:
+        # Returns the path to the new imagefile, or None if it already exists
+        imagefilename = f"{self.imagename}.sing"
+        imagefile = self.sourcefile.with_name(imagefilename)
+        if not imagefile.exists() and not imagefile.is_symlink():
+            # eventually can use docker builds, e.g.
+            # docker run -it -v $PWD:$PWD -w $PWD --privileged quay.io/singularity/singularity:v3.5.1 build ./busybox.sing-3.5.1 ./Singularity.busybox
+            # docker run -it -v $PWD:$PWD -w $PWD --entrypoint chown quay.io/singularity/singularity:v2.6 -- $UID.$GID ./busybox.sing
+            # unfortunately 2.6 version tries to connect to docker service and
+            # fails, thus can't fetch anything
+            tmpfile = self.sourcefile.with_name(f"{imagefilename}.tmp")
+            builder.runcmd(
+                "sudo",
+                "--preserve-env=TMPDIR,SINGULARITY_TMPDIR",
+                "singularity",
+                "build",
+                str(tmpfile),
+                str(self.sourcefile),
+            )
+            # abuse BTRFS CoW to avoid sudo/docker to chown
+            builder.runcmd("cp", "--reflink=auto", str(tmpfile), str(imagefile))
+            tmpfile.unlink()
+            return imagefile
+        else:
+            return None
+
+
+def version_key(vstr: str) -> tuple[tuple[int, ...], str]:
+    m = re.match(r"[0-9]+(?:\.[0-9]+)*", vstr)
+    assert m is not None
+    return (tuple(int(c) for c in m[0].split(".")), vstr[m.end() :])
+
+
+def get_familyname(repoid: str, family: str) -> str:
+    name = re.sub(r"^[^/]*/", "", repoid)
+    # sanitize for datalad not allowing _ in the container names
+    name = name.replace("_", "-")
+    return f"{family}-{name}"
+
+
+def get_imagename(repoid: str, family: str, versiontag: str) -> str:
+    familyname = get_familyname(repoid, family)
+    return f"{familyname}--{versiontag}"
+
+
+def main() -> None:
+    logging.basicConfig(format="[%(levelname)-8s] %(message)s", level=logging.INFO)
+
+    # Use local dedicated "tmp/" directory since any other might be too small
+    topdir = Path(__file__).parent.parent
+    topdir.mkdir(parents=True, exist_ok=True)
+    builder = Builder(
+        image_dir=topdir / "images",
```

Comment:
you can try running it until it fails on singularity call which should happen only if a new one to be added... which should not happen if this branch is based on recent version since unlikely a new image to be added.

if you try to run it you will get an error here due to typo that it should be `images_dir`.  `mypy` picks it up too

```
$> mypy scripts/create_singularities         
scripts/create_singularities:83: error: Argument "key" to "max" has incompatible type "Callable[[str], Tuple[Tuple[int, ...], str]]"; expected "Callable[[Optional[str]], Tuple[Tuple[int, ...], str]]"
scripts/create_singularities:250: error: Unexpected keyword argument "image_dir" for "Builder"; did you mean "images_dir"?
scripts/create_singularities:269: error: Incompatible types in assignment (expression has type "Response", variable has type "CompletedProcess[str]")
scripts/create_singularities:272: error: "CompletedProcess[str]" has no attribute "raise_for_status"
scripts/create_singularities:273: error: "CompletedProcess[str]" has no attribute "text"
Found 5 errors in 1 file (checked 1 source file)
```

Reply:
I fixed that issue, but when I run the code, it generates a singularity recipe for `bids/aa:latest`, and then `datalad push --data auto` fails with:

```
CommandError: 'git -c diff.ignoreSubmodules=none annex sync origin --no-push --no-pull --no-commit --no-resolvemerge --no-content -c annex.dotfiles=true' failed with exitcode 1 under /Users/jwodder/dartmouth/containers
git-annex: /Users/jwodder/dartmouth/containers/tmp: openTempFile template git-annex.tmp: does not exist (No such file or directory)
```

Reply:
it should not consider `latest` tag as we should filter only for actual versioned (just stripping some `v` prefix etc) as was done here: https://github.com/ReproNim/containers/blob/master/scripts/create_singularities#L106 ... may be there should be no `not` in your condition at https://github.com/ReproNim/containers/pull/75/files#diff-29ef84f03b68174af0e117047ee7d55f0791ec646287f827529418f30533fd3dR78 ?  not yet sure about what I meant with that matching for `master`...

Reply:
@yarikoptic Fixed, but now it's creating a Singularity recipe for `bids/cpac:v1.1.0_8`, and `datalad push` fails as before.

Reply:
interesting... we already have

```
$> ls images/bids/*cpac*
images/bids/Singularity.bids-cpac--1.1.0_14  images/bids/bids-cpac--1.1.0_14.sing
```

... I guess there is some difference in logic between bash and your python versions... looking at it...

Reply:
in your `version_key` you take only portion `.`-separated, here we need to take also the one after `_`.   I would take all `_` and `.` separated elements. Also I would not ignore string portions:
- do not match (previous regex application should strip away anything we do not want)
- split based on `.` or `_`
- convert to tuple of elements doing `int()` if `.isdigit` and keep str otherwise

Reply:
> convert to tuple of elements doing `int()` if `.isdigit` and keep str otherwise

I don't think that'll work; values of different types can't be ordered in Python 3.

Reply:
Pragmatically I hope we wouldn't hit comparison of different types... But ok, do only for digit components for now unless we run into use case for it. Or can use that https://pypi.org/project/looseversion/

Reply:
@yarikoptic So what exactly should happen to non-digit components?  Should they just be discarded?

Reply:
@yarikoptic I did that, but then when I ran the script again, it failed to sort the list `['0.3.1@v0.3.1', '0.3@v0.3', '0.2.2.2@v0.2.2.2', '0.2.2.1@v0.2.2.
1', '0.2.2@v0.2.2', '0.2.1.1-2@v0.2.1.1-2', '0.2.1.1-1@v0.2.1.1-1', '0.2-1@v0.2-
1']` with the error:

```
Traceback (most recent call last):
  File "/Users/jwodder/dartmouth/containers/scripts/create_singularities", line 316, in <module>
    main()
  File "/Users/jwodder/dartmouth/containers/scripts/create_singularities", line 284, in main
    builder.generate_singularity_for_docker_image(dockerhubid, "bids")
  File "/Users/jwodder/dartmouth/containers/scripts/create_singularities", line 190, in generate_singularity_for_docker_image
    last_version = self.get_last_docker_version_tag(dockerhubid)
  File "/Users/jwodder/dartmouth/containers/scripts/create_singularities", line 88, in get_last_docker_version_tag
    return max(good_versions, key=LooseVersion)
  File "/Users/jwodder/dartmouth/containers/scripts/.nox/create_singularities/lib/python3.10/site-packages/looseversion.py", line 147, in __gt__
    c = self._cmp(other)
  File "/Users/jwodder/dartmouth/containers/scripts/.nox/create_singularities/lib/python3.10/site-packages/looseversion.py", line 185, in _cmp
    if self.version < other.version:
TypeError: '<' not supported between instances of 'str' and 'int'
```

Reply:
all that `@` joining of "filtered" and original tag was to mitigate shortcoming (or my ignorance) of bash. In Python - just make a dict from filtered to original version, sort only versions (what is before `@`) and then get docker tag from the mapping.  I think what happens here now is that it gets that `@v` and tries to compare to e.g. `1` from previous one.

Reply:
@yarikoptic Now it's generating a recipe for bids/magetbrain:v0.3.1, and the `datalad push` error is still occurring.

Reply:
0.3.1 sounds like a proper version to upgrade current 0.3 to. You might have hit the bug of the current `bash` "poor man" version comparison + mapping solution I had.  I will give Python version a shot again now. Thanks!